x-pack/plugin/snapshot-repo-test-kit/src/main/java/org/elasticsearch/repositories/blobstore/testkit/RepositoryAnalyzeAction.java

```diff
@@ -661,6 +686,59 @@ public void onFailure(Exception exp) {
             };
         }
 
+        private class UncontendedRegisterAnalysis implements Runnable {
+            private final Random random;
+            private final String registerName;
+            private final List<DiscoveryNode> nodes;
+            private final AtomicBoolean otherAnalysisComplete;
+            private int currentValue; // actions run in strict sequence so no need for synchronization
```

Comment:
I think this isn't the case, but it's a good observation. If that were the case then we potentially would need synchronization here indeed. Fortunately for remote requests the request and response go over the same TCP channel which means they use the same transport worker thread ([docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/modules-network.html#modules-network-threading-model)) and therefore the request handling happens-before the response handling in program order on that thread so it's ok. Local requests bypass the transport worker threads of course, but they all happen within the same JVM so we have a proper happens-before relationship there too.

Reply:
Thanks. You are right. In previous convesations, I heard that we could in theory send request via one channel but receive the response via another since transport does not have the ordered processing constraint of HTTP 1.1. It is a theoretical possbility, not what we have today. Sorry that I mis-remembered.

Reply:
Technically speaking I think there may be no happens-before relationship between sending a request A and receiving a different request B which was caused by the remote node's handling of request A, because those things will use different TCP channels for sure and therefore may land on different transport threads. We do use nested requests in various places, e.g. recovery. I'm not sure if this is something that can happen in practice, but definitely something to watch out for.